src/Storages/MergeTree/MergeTreeSource.cpp

```diff
@@ -0,0 +1,220 @@
+        return control;
+    }
+
+    void schedule(ThreadPool::Job job)
+    {
+        callback_runner(std::move(job), 0);
+    }
+
+    ChunkAndProgress getResult()
+    {
+        return control->getResult();
+    }
+
+    Stage getStage() const { return control->stage; }
+    int getFD() const { return control->event.fd; }
+
+    AsyncReadingState()
+    {
+        control = std::make_shared<Control>();
+        callback_runner = threadPoolCallbackRunner<void>(IOThreadPool::get(), "MergeTreeRead");
+    }
+
+    ~AsyncReadingState()
+    {
+        /// Here we wait for async task if needed.
+        /// ~AsyncReadingState and Control::finish can be run concurrently.
+        /// It's important to store std::shared_ptr<Control> into bg pool task.
+        /// Otherwise following is possible:
+        ///
+        ///  (executing thread)                         (bg pool thread)
+        ///                                             Control::finish()
+        ///                                             stage = Stage::IsFinished;
+        ///  ~MergeTreeBaseSelectProcessor()
+        ///  ~AsyncReadingState()
+        ///  control->stage != Stage::InProgress
+        ///  ~EventFD()
+        ///                                             event.write()
+        if (control->stage == Stage::InProgress)
+            control->event.read();
+    }
+
+private:
+    ThreadPoolCallbackRunner<void> callback_runner;
+    std::shared_ptr<Control> control;
+};
+#endif
+
+ISource::Status MergeTreeSource::prepare()
+{
+#if defined(OS_LINUX)
+    if (!async_reading_state)
+        return ISource::prepare();
+
+    /// Check if query was cancelled before returning Async status. Otherwise it may lead to infinite loop.
+    if (isCancelled())
+    {
+        getPort().finish();
+        return ISource::Status::Finished;
+    }
+
+    if (async_reading_state && async_reading_state->getStage() == AsyncReadingState::Stage::InProgress)
+        return ISource::Status::Async;
+#endif
+
+    return ISource::prepare();
+}
+
+
+std::optional<Chunk> MergeTreeSource::reportProgress(ChunkAndProgress chunk)
+{
+    if (chunk.num_read_rows || chunk.num_read_bytes)
+        progress(chunk.num_read_rows, chunk.num_read_bytes);
+
+    if (chunk.chunk.hasRows())
+        return std::move(chunk.chunk);
+
+    return {};
+}
+
+
+std::optional<Chunk> MergeTreeSource::tryGenerate()
+{
+#if defined(OS_LINUX)
+    if (async_reading_state)
+    {
+        if (async_reading_state->getStage() == AsyncReadingState::Stage::IsFinished)
+            return reportProgress(async_reading_state->getResult());
+
+        chassert(async_reading_state->getStage() == AsyncReadingState::Stage::NotStarted);
+
+        /// It is important to store control into job.
+        /// Otherwise, race between job and ~MergeTreeBaseSelectProcessor is possible.
+        auto job = [this, control = async_reading_state->start()]() mutable
+        {
+            auto holder = std::move(control);
```

Comment:
do we really need this object?

Reply:
Actually no. In our thread pool implementation we set `job = {}` after execution.
But I still prefer to do it just in case. So that `holder` is removed after functions finishes. And not stored in lambda capture till lambda is destroyed.